lib/std/fs/path.zig

```diff
@@ -1318,3 +1333,552 @@ test "stem" {
+
+        pub fn last(self: *Self) ?Component {
+            self.end_index = self.path.len;
+            while (true) {
+                if (self.end_index == self.root_end_index) {
+                    self.start_index = self.end_index;
+                    return null;
+                }
+                if (!path_type.isSep(T, self.path[self.end_index - 1])) break;
+                self.end_index -= 1;
+            }
+            self.start_index = self.end_index;
+            while (true) {
+                if (self.start_index == self.root_end_index) break;
+                if (path_type.isSep(T, self.path[self.start_index - 1])) break;
+                self.start_index -= 1;
+            }
+            if (self.start_index == self.end_index) return null;
+            return .{
+                .name = self.path[self.start_index..self.end_index],
+                .path = self.path[0..self.end_index],
+            };
+        }
+
+        pub fn next(self: *Self) ?Component {
+            var start_index = self.end_index;
+            while (start_index < self.path.len and path_type.isSep(T, self.path[start_index])) {
+                start_index += 1;
+            }
+            var end_index = start_index;
+            while (end_index < self.path.len and !path_type.isSep(T, self.path[end_index])) {
+                end_index += 1;
+            }
+            if (start_index == end_index) return null;
+            self.start_index = start_index;
+            self.end_index = end_index;
+            return .{
+                .name = self.path[self.start_index..self.end_index],
+                .path = self.path[0..self.end_index],
+            };
+        }
+
+        pub fn previous(self: *Self) ?Component {
+            var end_index = self.start_index;
+            while (true) {
+                if (end_index == self.root_end_index) return null;
+                if (!path_type.isSep(T, self.path[end_index - 1])) break;
+                end_index -= 1;
+            }
+            var start_index = end_index;
+            while (true) {
+                if (start_index == self.root_end_index) break;
+                if (path_type.isSep(T, self.path[start_index - 1])) break;
+                start_index -= 1;
+            }
+            if (start_index == end_index) return null;
+            self.start_index = start_index;
+            self.end_index = end_index;
+            return .{
+                .name = self.path[self.start_index..self.end_index],
+                .path = self.path[0..self.end_index],
+            };
+        }
+    };
+}
+
+pub const NativeUtf8ComponentIterator = ComponentIterator(switch (native_os) {
+    .windows => .windows,
+    .uefi => .uefi,
+    else => .posix,
+}, u8);
+
+pub fn componentIterator(path: []const u8) !NativeUtf8ComponentIterator {
+    return NativeUtf8ComponentIterator.init(path);
+}
+
+test "ComponentIterator posix" {
+    const PosixComponentIterator = ComponentIterator(.posix, u8);
```

Comment:
What about utf-8 paths and relative paths?

Reply:
Relative paths are tested. UTF-8 paths don't need any special handling to work correctly in the implementation so a test case with non-ASCII UTF-8 codepoints wouldn't really test anything new.